### PR TITLE
[MINOR][DOCS] JSON APIs related documentation fixes

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes corrections related to JSON APIs as below:

- Rendering links in Python documentation
- Replacing `RDD` to `Dataset` in programing guide
- Adding missing description about JSON Lines consistently in `DataFrameReader.json` in Python API
- De-duplicating little bit of `DataFrameReader.json` in Scala/Java API 

## How was this patch tested?

Manually build the documentation via `jekyll build`. Corresponding snapstops will be left on the codes.

Note that currently there are Javadoc8 breaks in several places. These are proposed to be handled in https://github.com/apache/spark/pull/17477. So, this PR does not fix those.

### CODE REVIEW

File: docs/sql-programming-guide.md [hunk 1]
```diff
@@ -883,7 +883,7 @@ Configuration of Parquet can be done using the `setConf` method on `SparkSession
 
 <div data-lang="scala"  markdown="1">
 Spark SQL can automatically infer the schema of a JSON dataset and load it as a `Dataset[Row]`.
-This conversion can be done using `SparkSession.read.json()` on either an RDD of String,
+This conversion can be done using `SparkSession.read.json()` on either an Dataset of String,
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Scala example uses `Dataset` as below:

File: docs/sql-programming-guide.md [hunk 2]
```diff
@@ -897,7 +897,7 @@ For a regular multi-line JSON file, set the `wholeFile` option to `true`.
 
 <div data-lang="java"  markdown="1">
 Spark SQL can automatically infer the schema of a JSON dataset and load it as a `Dataset<Row>`.
-This conversion can be done using `SparkSession.read().json()` on either an RDD of String,
+This conversion can be done using `SparkSession.read().json()` on either an Dataset of String,
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Java example uses `Dataset` as below:

File: python/pyspark/sql/readwriter.py [hunk 3]
```diff
@@ -173,8 +173,8 @@ def json(self, path, schema=None, primitivesAsString=None, prefersDecimal=None,
         """
         Loads JSON files and returns the results as a :class:`DataFrame`.
 
-        `JSON Lines <http://jsonlines.org/>`_(newline-delimited JSON) is supported by default.
-        For JSON (one record per file), set the `wholeFile` parameter to ``true``.
+        `JSON Lines <http://jsonlines.org/>`_ (newline-delimited JSON) is supported by default.
+        For JSON (one record per file), set the ``wholeFile`` parameter to ``true``.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: **Before**




**After**

File: python/pyspark/sql/readwriter.py [hunk 4]
```diff
@@ -634,7 +634,9 @@ def saveAsTable(self, name, format=None, mode=None, partitionBy=None, **options)
 
     @since(1.4)
     def json(self, path, mode=None, compression=None, dateFormat=None, timestampFormat=None):
-        """Saves the content of the :class:`DataFrame` in JSON format at the specified path.
+        """Saves the content of the :class:`DataFrame` in JSON format
+        (`JSON Lines text format or newline-delimited JSON <[http://jsonlines.org/>`_) at the
+        specified path.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: **Before **


**After**



Note that this is not consistent with Scala/Java ones:

File: python/pyspark/sql/streaming.py [hunk 5]
```diff
@@ -405,8 +405,8 @@ def json(self, path, schema=None, primitivesAsString=None, prefersDecimal=None,
         """
         Loads a JSON file stream and returns the results as a :class:`DataFrame`.
 
-        `JSON Lines <http://jsonlines.org/>`_(newline-delimited JSON) is supported by default.
-        For JSON (one record per file), set the `wholeFile` parameter to ``true``.
+        `JSON Lines <http://jsonlines.org/>`_ (newline-delimited JSON) is supported by default.
+        For JSON (one record per file), set the ``wholeFile`` parameter to ``true``.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/streaming.py]: **Before**



**After**

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 6]
```diff
@@ -268,8 +268,8 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
   }
 
   /**
-   * Loads a JSON file (<a href="http://jsonlines.org/">JSON Lines text format or
-   * newline-delimited JSON</a>) and returns the result as a `DataFrame`.
+   * Loads a JSON file and returns the results as a `DataFrame`.
+   *
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: This de-duplicate the documentation as it points the overloaded `json()` out below.

**Before**




**After**

File: docs/sql-programming-guide.md [hunk 7]
```diff
@@ -883,7 +883,7 @@ Configuration of Parquet can be done using the `setConf` method on `SparkSession
 
 <div data-lang="scala"  markdown="1">
 Spark SQL can automatically infer the schema of a JSON dataset and load it as a `Dataset[Row]`.
-This conversion can be done using `SparkSession.read.json()` on either an RDD of String,
+This conversion can be done using `SparkSession.read.json()` on either an `Dataset[String]`,
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Scala example uses `Dataset` as below:

File: python/pyspark/sql/readwriter.py [hunk 8]
```diff
@@ -634,7 +634,9 @@ def saveAsTable(self, name, format=None, mode=None, partitionBy=None, **options)
 
     @since(1.4)
     def json(self, path, mode=None, compression=None, dateFormat=None, timestampFormat=None):
-        """Saves the content of the :class:`DataFrame` in JSON format at the specified path.
+        """Saves the content of the :class:`DataFrame` in JSON format
+        (`JSON Lines text format or newline-delimited JSON <http://jsonlines.org/>`_) at the
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: **Before**


**After**



Note that this is not consistent with Scala/Java ones:

File: docs/sql-programming-guide.md [hunk 9]
```diff
@@ -897,7 +897,7 @@ For a regular multi-line JSON file, set the `wholeFile` option to `true`.
 
 <div data-lang="java"  markdown="1">
 Spark SQL can automatically infer the schema of a JSON dataset and load it as a `Dataset<Row>`.
-This conversion can be done using `SparkSession.read().json()` on either an RDD of String,
+This conversion can be done using `SparkSession.read().json()` on either an `Dataset<String>`,
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Java example uses `Dataset` as below:



Output:

File: docs/sql-programming-guide.md [hunk 10]
```diff
@@ -883,7 +883,7 @@ Configuration of Parquet can be done using the `setConf` method on `SparkSession
 
 <div data-lang="scala"  markdown="1">
 Spark SQL can automatically infer the schema of a JSON dataset and load it as a `Dataset[Row]`.
-This conversion can be done using `SparkSession.read.json()` on either an RDD of String,
+This conversion can be done using `SparkSession.read.json()` on either a `Dataset[String]`,
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Output:



Example:

File: docs/sql-programming-guide.md [hunk 11]
```diff
@@ -897,7 +897,7 @@ For a regular multi-line JSON file, set the `wholeFile` option to `true`.
 
 <div data-lang="java"  markdown="1">
 Spark SQL can automatically infer the schema of a JSON dataset and load it as a `Dataset<Row>`.
-This conversion can be done using `SparkSession.read().json()` on either an RDD of String,
+This conversion can be done using `SparkSession.read().json()` on either a `Dataset<String>`,
```
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Output:


Example:

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by HyukjinKwon** [comment]
Thank you @srowen.

**Pull request closed without merging**